src/libostree/ostree-repo-prune.c

```diff
@@ -46,18 +46,43 @@ maybe_prune_loose_object (OtPruneData        *data,
                           GCancellable       *cancellable,
                           GError            **error)
 {
+  gboolean reachable = FALSE;
   g_autoptr(GVariant) key = NULL;
 
   key = ostree_object_name_serialize (checksum, objtype);
 
-  if (!g_hash_table_lookup_extended (data->reachable, key, NULL, NULL))
+  if (g_hash_table_lookup_extended (data->reachable, key, NULL, NULL))
+    reachable = TRUE;
+  else
     {
       g_debug ("Pruning unneeded object %s.%s", checksum,
                ostree_object_type_to_string (objtype));
       if (!(flags & OSTREE_REPO_PRUNE_FLAGS_NO_PRUNE))
         {
           guint64 storage_size = 0;
 
+          if (objtype == OSTREE_OBJECT_TYPE_PAYLOAD_LINK)
+            {
+              char loose_path_buf[_OSTREE_LOOSE_PATH_MAX];
+              char target_checksum[OSTREE_SHA256_STRING_LEN+1];
+
+              _ostree_loose_path (loose_path_buf, checksum, OSTREE_OBJECT_TYPE_PAYLOAD_LINK, data->repo->mode);
+              if (readlinkat (data->repo->objects_dir_fd, loose_path_buf, target_checksum, sizeof (target_checksum)) < 0)
+                return glnx_throw_errno (error);
+
+              g_autoptr(GVariant) target_key = ostree_object_name_serialize (target_checksum, OSTREE_OBJECT_TYPE_FILE);
+
+              if (g_hash_table_lookup_extended (data->reachable, target_key, NULL, NULL))
+                {
+                  if (!ostree_repo_query_object_storage_size (data->repo, OSTREE_OBJECT_TYPE_FILE, target_checksum,
```

Comment:
Should also handle `ENOENT` here right?

Reply:
hm.. in that case it won't be in the `data->reachable` hash table?

Reply:
I've addressed all the other comments, except this one and checking for the reflink support in the file system